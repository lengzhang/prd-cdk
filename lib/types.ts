import * as cdk from "aws-cdk-lib";

export interface Environment extends cdk.Environment {
  stage: string;
}
