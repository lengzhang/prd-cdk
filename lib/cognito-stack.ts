import * as cdk from "aws-cdk-lib";
import * as cognito from "aws-cdk-lib/aws-cognito";
import * as cognitoIdentityPool from "@aws-cdk/aws-cognito-identitypool-alpha";
import * as iam from "aws-cdk-lib/aws-iam";
import { Construct } from "constructs";

interface CognitoStackProps extends cdk.StackProps {
  userPoolName: string;
  userPoolClientName: string;
  userPoolDomainName: string;
  userPoolDomainPrefix: string;
  identityPoolName: string;
  reportsBucketName: string;
}

class CognitoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CognitoStackProps) {
    super(scope, id, props);

    const { userPool, userPoolClient } = this.createUserPool(
      props.userPoolName,
      props.userPoolClientName,
      props.userPoolDomainName,
      props.userPoolDomainPrefix
    );

    this.createIdentityPool(
      props.identityPoolName,
      userPool,
      userPoolClient,
      props.reportsBucketName
    );
  }

  createUserPool(
    userPoolName: string,
    userPoolClientName: string,
    userPoolDomainName: string,
    userPoolDomainPrefix: string
  ) {
    const userVerification: cognito.UserVerificationConfig = {
      emailSubject: "[PRD] You need to verify your email",
      emailBody:
        "Thanks for signing up. Please verify your account by the verification code {####}.",
      emailStyle: cognito.VerificationEmailStyle.CODE,
    };
    const standardAttributes: cognito.StandardAttributes = {
      email: { mutable: false, required: true },
      preferredUsername: { mutable: true, required: true },
      lastUpdateTime: { mutable: false, required: true },
    };

    const customAttributes: {
      [key: string]: cdk.aws_cognito.ICustomAttribute;
    } = {
      // timestamp for date time
      created_at: new cognito.NumberAttribute({ mutable: false }),
    };

    const passwordPolicy: cognito.PasswordPolicy = {
      minLength: 8,
      requireLowercase: true,
      requireUppercase: true,
      requireDigits: true,
      requireSymbols: false,
    };

    const userPool = new cognito.UserPool(this, userPoolName, {
      userPoolName,
      signInAliases: { email: true },
      selfSignUpEnabled: true,
      autoVerify: { email: true },
      userVerification,
      standardAttributes,
      customAttributes,
      passwordPolicy,
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    userPool.addDomain(userPoolDomainName, {
      cognitoDomain: {
        domainPrefix: userPoolDomainPrefix,
      },
    });

    const userPoolClient = userPool.addClient(userPoolClientName, {
      userPoolClientName: userPoolClientName,
      authFlows: {
        userPassword: true,
      },
    });

    return { userPool, userPoolClient };
  }

  createIdentityPool(
    identityPoolName: string,
    userPool: cognito.UserPool,
    userPoolClient: cognito.UserPoolClient,
    reportBucketName: string
  ) {
    const userPoolAuthenticationProvider =
      new cognitoIdentityPool.UserPoolAuthenticationProvider({
        userPool,
        userPoolClient,
      });
    const identityPool = new cognitoIdentityPool.IdentityPool(
      this,
      identityPoolName,
      { identityPoolName }
    );

    identityPool.addUserPoolAuthentication(userPoolAuthenticationProvider);
    identityPool.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);

    const reportBucketAccessPolicy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["s3:GetObject", "s3:ListBucket", "s3:PutObject"],
      resources: [
        `arn:aws:s3:::${reportBucketName}`,
        `arn:aws:s3:::${reportBucketName}/*`,
      ],
    });
    identityPool.authenticatedRole.addToPrincipalPolicy(
      reportBucketAccessPolicy
    );
  }
}

export default CognitoStack;
