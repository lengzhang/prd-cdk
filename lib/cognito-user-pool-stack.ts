import * as cdk from "aws-cdk-lib";
import * as cognito from "aws-cdk-lib/aws-cognito";
import { Construct } from "constructs";

interface CognitoUserPoolStackProps extends cdk.StackProps {
  userPoolName: string;
  appClientName: string;
  cognitoDomainName: string;
  cognitoDomainPrefix: string;
}

class CognitoUserPoolStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CognitoUserPoolStackProps) {
    super(scope, id, props);

    this.createUserPool(
      props.userPoolName,
      props.appClientName,
      props.cognitoDomainName,
      props.cognitoDomainPrefix
    );
  }

  createUserPool(
    userPoolName: string,
    appClientName: string,
    cognitoDomainName: string,
    cognitoDomainPrefix: string
  ) {
    const userVerification: cognito.UserVerificationConfig = {
      emailSubject: "[PRD] You need to verify your email",
      emailBody:
        "Thanks for signing up. Please verify your account by the verification code {####}.",
      emailStyle: cognito.VerificationEmailStyle.CODE,
    };
    const standardAttributes: cognito.StandardAttributes = {
      email: { mutable: false, required: true },
      familyName: { mutable: true, required: true },
      givenName: { mutable: true, required: true },
      preferredUsername: { mutable: true, required: true },
      lastUpdateTime: { mutable: false, required: true },
    };

    const customAttributes: {
      [key: string]: cdk.aws_cognito.ICustomAttribute;
    } = {
      // timestamp for date time
      created_at: new cognito.NumberAttribute({ mutable: false }),
    };

    const passwordPolicy: cognito.PasswordPolicy = {
      minLength: 8,
      requireLowercase: true,
      requireUppercase: true,
      requireDigits: true,
      requireSymbols: false,
    };

    const userPool = new cognito.UserPool(this, userPoolName, {
      userPoolName,
      signInAliases: { email: true },
      selfSignUpEnabled: true,
      autoVerify: { email: true },
      userVerification,
      standardAttributes,
      customAttributes,
      passwordPolicy,
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    userPool.addDomain(cognitoDomainName, {
      cognitoDomain: {
        domainPrefix: cognitoDomainPrefix,
      },
    });

    userPool.addClient(appClientName, {
      userPoolClientName: appClientName,
      authFlows: {
        userPassword: true,
      },
    });
  }
}

export default CognitoUserPoolStack;
