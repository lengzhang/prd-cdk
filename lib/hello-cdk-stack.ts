import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

interface HelloCdkStackProps extends cdk.StackProps {
  bucketName: string;
}

class HelloCdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: HelloCdkStackProps) {
    super(scope, id, props);

    new cdk.aws_s3.Bucket(this, props.bucketName, {
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
  }
}

export default HelloCdkStack;
