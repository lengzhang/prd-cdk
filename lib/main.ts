#!/usr/bin/env node
import "source-map-support/register";
import * as dotenv from "dotenv";

import * as cdk from "aws-cdk-lib";

import PrdResourcesStack from "./prd-resources-stack";
import CognitoUserPoolStack from "./cognito-stack";
import { Environment } from "./types";

dotenv.config();

const appName = "prd";
const env: Environment = {
  account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
  stage: process.env.STAGE || "",
};

function main() {
  const app = new cdk.App();
  const namePrefix = `${env.stage}-${appName}-${env.account}-${env.region}`;

  const prdResourcesStackName = `${namePrefix}-resources-stack`;
  const prdResourcesBucketName = `${namePrefix}-reports-bucket`;
  new PrdResourcesStack(app, prdResourcesStackName, {
    env,
    reportsBucketName: prdResourcesBucketName,
  });

  const cognitoStackName = `${namePrefix}-cognito-stack`;
  const userPoolName = `${namePrefix}-cognito-user-pool`;
  const userPoolClientName = `${namePrefix}-cognito-user-pool-client`;
  const userPoolDomainName = `${namePrefix}-cognito-user-pool-domain`;
  const userPoolDomainPrefix = `${env.stage}-${appName}-${env.account}`;
  const identityPoolName = `${namePrefix}-cognito-identity-pool`;
  new CognitoUserPoolStack(app, cognitoStackName, {
    env,
    userPoolName,
    userPoolClientName,
    userPoolDomainName,
    userPoolDomainPrefix,
    identityPoolName,
    reportsBucketName: prdResourcesBucketName,
  });
}

main();
