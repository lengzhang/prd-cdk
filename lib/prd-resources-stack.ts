import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

interface PrdResourcesStackProps extends cdk.StackProps {
  reportsBucketName: string;
}

class PrdResourcesStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: PrdResourcesStackProps) {
    super(scope, id, props);

    this.createPrdReportsBucket(props.reportsBucketName);
  }

  createPrdReportsBucket(reportsBucketName: string) {
    new cdk.aws_s3.Bucket(this, reportsBucketName, {
      bucketName: reportsBucketName,
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
  }
}

export default PrdResourcesStack;
