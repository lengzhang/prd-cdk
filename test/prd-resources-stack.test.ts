import * as cdk from "aws-cdk-lib";
import { Template, Match } from "aws-cdk-lib/assertions";

import PrdResourcesStack from "../lib/prd-resources-stack";
import { Environment } from "../lib/types";

const env: Environment = {
  account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
  stage: "test",
};

describe("PrdResourcesStack", () => {
  const getPrdResourcesStack = () => {
    const app = new cdk.App();

    const stack = new PrdResourcesStack(app, "test-stack-name", {
      env,
      reportsBucketName: "test-bucket-name",
    });

    return stack;
  };

  const stack = getPrdResourcesStack();
  const template = Template.fromStack(stack);

  test("matches the snapshot", () => {
    expect(stack.stackName).toBe("test-stack-name");
    expect(template.toJSON()).toMatchSnapshot();
  });

  test("can be auto deleted", () => {
    template.hasResource("AWS::S3::Bucket", {
      DeletionPolicy: "Delete",
    });
    template.hasResourceProperties("AWS::S3::Bucket", {
      Tags: Match.arrayEquals([
        { Key: "aws-cdk:auto-delete-objects", Value: "true" },
      ]),
    });
  });

  test("should have versioning", () => {
    template.hasResourceProperties("AWS::S3::Bucket", {
      VersioningConfiguration: Match.objectEquals({
        Status: "Enabled",
      }),
    });
  });
});
