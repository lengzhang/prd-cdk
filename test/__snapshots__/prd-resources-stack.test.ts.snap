// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrdResourcesStack matches the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b7f33614a69548d6bafe224d751a7ef238cde19097415e553fe8b63a4c8fd8a6.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "testbucketname086F697F",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testbucketname086F697F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": "test-bucket-name",
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "testbucketnameAutoDeleteObjectsCustomResource4751D172": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "testbucketnamePolicyB6C7B6DF",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "testbucketname086F697F",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "testbucketnamePolicyB6C7B6DF": {
      "Properties": {
        "Bucket": {
          "Ref": "testbucketname086F697F",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "testbucketname086F697F",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "testbucketname086F697F",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
